core/src/main/java/io/grpc/util/AdvancedTlsX509TrustManager.java

```diff
@@ -0,0 +1,399 @@
+      X509ExtendedTrustManager delegateManager = null;
+      TrustManagerFactory tmf;
+      try {
+        tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+        tmf.init(ks);
+      } catch (KeyStoreException | NoSuchAlgorithmException e) {
+        throw new CertificateException("Failed to create delegate TrustManagerFactory", e);
+      }
+      TrustManager[] tms = tmf.getTrustManagers();
+      // Iterate over the returned trust managers, looking for an instance of X509TrustManager.
+      // If found, use that as the delegate trust manager.
+      for (i = 0; i < tms.length; i++) {
+        if (tms[i] instanceof X509ExtendedTrustManager) {
+          delegateManager = (X509ExtendedTrustManager) tms[i];
+          break;
+        }
+      }
+      if (delegateManager == null) {
+        throw new CertificateException(
+            "Instance delegateX509TrustManager is null. Failed to initialize");
+      }
+      // Configure the delegated trust manager based on users' input.
+      if (checkingServer) {
+        if (this.verification == Verification.CertificateAndHostNameVerification
+            || this.verification == Verification.CertificateOnlyVerification) {
+          if (sslEngine == null) {
+            throw new CertificateException(
+                "SSLEngine is null. Couldn't check host name");
+          }
+          String algorithm = this.verification == Verification.CertificateAndHostNameVerification
+              ? "HTTPS" : "";
+          SSLParameters sslParams = sslEngine.getSSLParameters();
+          sslParams.setEndpointIdentificationAlgorithm(algorithm);
+          sslEngine.setSSLParameters(sslParams);
+        }
+        delegateManager.checkServerTrusted(chain, authType, sslEngine);
+      } else {
+        delegateManager.checkClientTrusted(chain, authType, sslEngine);
+      }
+    }
+    // Perform the additional peer cert check.
+    if (sslEngine != null && sslEnginePeerVerifier != null) {
+      sslEnginePeerVerifier.verifyPeerCertificate(chain, authType, sslEngine);
+    }
+    if (socket != null && sslSocketPeerVerifier != null) {
+      sslSocketPeerVerifier.verifyPeerCertificate(chain, authType, socket);
+    }
+    if (peerVerifier != null) {
+      peerVerifier.verifyPeerCertificate(chain, authType);
+    }
+    if (sslEnginePeerVerifier == null && sslSocketPeerVerifier == null && peerVerifier == null) {
+      log.log(Level.INFO, "No peer verifier is specified");
+    }
+  }
+
+  /**
+   * starts a {@code ScheduledExecutorService} to read trust certificates from a local file path
+   * periodically, and update the cached trust certs if there is an update.
+   *
+   * @param trustCertFile  the file on disk holding the trust certificates
+   * @param initialDelay the time to delay first read-and-update execution
+   * @param delay the period between successive read-and-update executions
+   * @param unit the time unit of the initialDelay and period parameters
```

Comment:
s/period/delay/ ? Or do you want to rename the delay parameter?

Reply:
Let's use `period` then, since it captures the meaning that this is a periodical service.